techniques/applications/repoGpgKeyManagement/1.0/repoGpgKeyManagement.st

```diff
@@ -0,0 +1,161 @@
+
+   repo_gpg_uses_rpm::
+    "repo_gpg_get_keys" string => "${paths.rpm} -q gpg-pubkey --qf '%{description}\n' | /usr/bin/gpg --with-colon --keyid-format=long",
+      comment => "We use GPG to decode the whole PGP PUBLIC KEY BLOCK because rpm only stores the first 8 chars of the hash, which is suspect to collision attacks";
+
+   repo_gpg_uses_apt::
+    "repo_gpg_get_keys" string => "${paths.apt_key} adv --fingerprint --with-colon --keyid-format=long",
+      comment => "'apt-key adv' is a cascaded invocation of gpg directly, so we can use the same parameters as for rpm and parse the same output format later";
+
+  classes:
+
+    "repo_gpg_key_${repo_gpg_index}_action_remove"
+      expression => strcmp("${repo_gpg_action[${repo_gpg_index}]}","del");
+
+    "repo_gpg_key_${repo_gpg_index}_action_add"
+      expression => strcmp("${repo_gpg_action[${repo_gpg_index}]}","add");
+
+    "repo_gpg_uses_rpm"
+      and => { isvariable("paths.rpm"),     isexecutable("${paths.rpm}") };
+
+    "repo_gpg_uses_apt"
+      and => { isvariable("paths.apt_key"), isexecutable("${paths.apt_key}") };
+
+    "repo_gpg_key_${repo_gpg_index}_has_last8_uc"
+      expression => regextract(".{8}(.{8})", "${repo_gpg_hash_uc[${repo_gpg_index}]}", "repo_gpg_hash_last8_uc[${repo_gpg_index}]"),
+      ifvarclass => "repo_gpg_uses_apt.repo_gpg_key_${repo_gpg_index}_action_remove",
+      comment    => "We need the last 8 UPPER-cased chars for apt-key del, it does not work with the full length key ID. Somebody should open them a ticket...";
+
+    "repo_gpg_key_${repo_gpg_index}_has_last8_lc"
+      expression => regextract(".{8}(.{8})", "${repo_gpg_hash_lc[${repo_gpg_index}]}", "repo_gpg_hash_last8_lc[${repo_gpg_index}]"),
+      ifvarclass => "repo_gpg_uses_rpm.repo_gpg_key_${repo_gpg_index}_action_remove",
+      comment    => "We also need the last 8 lower-cased chars for rpm -e for package removal, so we cut it down...";
+
+  files:
+
+    "${g.rudder_var_tmp}/repo_gpg_key_${repo_gpg_index}_content"
+      comment       => "Create a temporary file for repo to import the GPG key. Unfortunately rpm does not accept stdin or - for importing, so we need this temporary.",
+      create        => "true",
+      edit_line     => insert_lines( "${repo_gpg_content[${repo_gpg_index}]}" ),
+      edit_defaults => ncf_empty_select("true"),
+      ifvarclass    => "repo_gpg_hashes_read_ok.!repo_gpg_key_${repo_gpg_hash_lc[${repo_gpg_index}]}_present.repo_gpg_key_${repo_gpg_index}_action_add",
+      classes       => classes_generic("repo_gpg_key_${repo_gpg_index}_file_created");
+
+    "${g.rudder_var_tmp}/repo_gpg_key_${repo_gpg_index}_content"
+      comment    => "Remove temporary file if the GPG key import has been tried, regardless of the outcome",
+      delete     => tidy,
+      ifvarclass => "repo_gpg_key_${repo_gpg_index}_imported_reached";
```

Comment:
I could put it to /dev/shm, a ramdisk, the size of the files would be 2-4kBytes each for the short time of being imported. That should fit to memory on any system running deb/rpm packages and is supported to run a Rudder agent...